pkg/fqdn/cache.go

```diff
@@ -154,9 +162,56 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
 		c.forward[entry.Name] = entries
 	}
 	c.updateWithEntryIPs(entries, entry)
-	// When lookupTime is much earlier than time.Now(), we may not expire all
-	// entries that should be expired, leaving more work for .Lookup.
-	c.removeExpired(entries, time.Now())
+
+}
+
+// AddNameToCleanup adds the IP with the given TTL to the the cleanup map to
+// delete the entry from the policy when it expires.
+// Need to be called with the lock
+func (c *DNSCache) addNameToCleanup(entry *cacheEntry) {
+	expiration := int64(entry.TTL) + time.Now().Unix()
+	expiredEntries, exists := c.cleanup[expiration]
+	if !exists {
+		expiredEntries = []string{}
+	}
+	c.cleanup[expiration] = append(expiredEntries, entry.Name)
+}
+
+// CleanupExpiredEntries cleans all the expired entries from the lastTime that
+// runs to the give time. It will lock the struct until retrieves all the data.
+// It returns the list of names that need to be deleted from the policies.
+func (c *DNSCache) CleanupExpiredEntries(expires time.Time) ([]string, time.Time) {
+	c.Lock()
+	defer c.Unlock()
+	timediff := int(expires.Sub(c.lastCleanup).Seconds())
+	expiredEntries := []string{}
+
+	for i := 0; i < int(timediff); i++ {
+		expiredTime := c.lastCleanup.Add(time.Second)
+		key := expiredTime.Unix()
+		c.lastCleanup = expiredTime
+		entries, exists := c.cleanup[key]
+		if !exists {
+			continue
+		}
+		expiredEntries = append(expiredEntries, entries...)
+		delete(c.cleanup, key)
+	}
+	uniqMap := make(map[string]bool)
+	for _, v := range expiredEntries {
+		uniqMap[v] = true
+	}
+
+	uniqExpires := make([]string, 0, len(uniqMap))
+	for v := range uniqMap {
+		uniqExpires = append(uniqExpires, v)
+	}
+	for _, name := range uniqExpires {
+		if entries, exists := c.forward[name]; exists {
+			c.removeExpired(entries, c.lastCleanup)
+		}
+	}
+	return uniqExpires, c.lastCleanup
```

Comment:
Oh, I meant the original lastCleanup at the start of this function, not the final one (which is going to be roughly `expires`).